src/Authoring/cswinmd/Program.cs

```diff
@@ -0,0 +1,137 @@
+
+            string inputText = File.ReadAllText(inputFile);
+
+            string componentName = Path.GetFileNameWithoutExtension(inputFile);
+
+            var assemblyName = componentName;
+
+            var windows_winmd = GetWindowsWinMdPath(sdkVersion);
+            var compilation = CSharpCompilation.Create(
+                assemblyName: componentName,
+                syntaxTrees: new[] { CSharpSyntaxTree.ParseText(inputText, new CSharpParseOptions(LanguageVersion.Preview), inputFile) },
+                references: new[] { MetadataReference.CreateFromFile(windows_winmd),
+                MetadataReference.CreateFromFile(typeof(Binder).Assembly.Location)
+                },
+                options: new CSharpCompilationOptions(OutputKind.DynamicallyLinkedLibrary)
+                );
+
+            var g = new SourceGenerator();
+            var outFolder = string.IsNullOrEmpty(o) ? Environment.GetEnvironmentVariable("TEMP") : o;
+            if (!Directory.Exists(outFolder))
+            {
+                Directory.CreateDirectory(outFolder);
+            }
+            try
+            {
+                var cp = new ConfigProvider();
+                var config = cp.GlobalOptions as ConfigOptions;
+                config.Values["build_property.AssemblyName"] = assemblyName;
+                config.Values["build_property.AssemblyVersion"] = "0.0.0.1";
+                config.Values["build_property.CsWinRTGeneratedFilesDir"] = outFolder;
+                config.Values["build_property.CsWinRTEnableLogging"] = "true";
+                config.Values["build_property.CsWinRTGenerateWinMDOnly"] = "true";
+
+                var driver = CSharpGeneratorDriver.Create(
+                    generators: ImmutableArray.Create(g),
+                    additionalTexts: ImmutableArray<AdditionalText>.Empty,
+                    parseOptions: (CSharpParseOptions)compilation.SyntaxTrees.First().Options,
+                    optionsProvider: cp
+                    );
+
+                var d = driver.RunGenerators(compilation);
+
+                var res = d.GetRunResult();
+                if (!res.Diagnostics.IsEmpty)
+                {
+                    foreach (var v in res.Diagnostics)
+                    {
+                        Console.WriteLine(v.GetMessage());
+                        Console.WriteLine($"\t\tIn {v.Location.GetLineSpan()}");
+                    }
+                    Console.WriteLine();
+                }
+            }
+            catch (Exception e)
+            {
+                Console.WriteLine(File.ReadAllText(Path.Join(outFolder, "log.txt")));
+            }
+        }
+
+        private static bool IsVersion(string v)
+        {
+            return Version.TryParse(Path.GetFileName(v), out var _);
+        }
+        private static string GetWindowsWinMdPath(string? sdkVersion)
+        {
+            using (var hklm = RegistryKey.OpenBaseKey(RegistryHive.LocalMachine, RegistryView.Registry32))

```

Comment:
We can reduce nesting even further:

```cs
using var hklm = RegistryKey.OpenBaseKey(RegistryHive.LocalMachine, RegistryView.Registry32);
using var roots = hklm.OpenSubKey(@"SOFTWARE\Microsoft\Windows Kits\Installed Roots");

var kitsRoot10 = (string)roots.GetValue("KitsRoot10");
var unionMetadata = Path.Combine(kitsRoot10, "UnionMetadata");
if (sdkVersion == null)
{
    var dirs = Directory.EnumerateDirectories(unionMetadata);
    sdkVersion = Path.GetFileName(dirs.Where(IsVersion).Last());
}

var path = Path.Combine(kitsRoot10, "UnionMetadata", sdkVersion, "Windows.winmd");
return path;
```

Not sure why the throw is here, it shouldn't be hit because there is a return right before it.